src/appleseed/renderer/modeling/bssrdf/randomwalkbssrdf.cpp

```diff
@@ -0,0 +1,408 @@
+            m_brdf_data.m_reflectance.set(1.0f);
+            m_brdf_data.m_reflectance_multiplier = 1.0f;
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override
+        {
+            return Model;
+        }
+
+        size_t compute_input_data_size() const override
+        {
+            return sizeof(RandomWalkBSSRDFInputValues);
+        }
+
+        void prepare_inputs(
+            Arena&                  arena,
+            const ShadingPoint&     shading_point,
+            void*                   data) const override
+        {
+            RandomWalkBSSRDFInputValues* values =
+                static_cast<RandomWalkBSSRDFInputValues*>(data);
+
+            new (&values->m_volume_data) GenericVolumeInputValues();
+
+            // Apply multipliers to input values.
+            values->m_reflectance *= values->m_reflectance_multiplier;
+            values->m_mfp *= values->m_mfp_multiplier;
+
+            // Clamp input values.
+            clamp_in_place(values->m_reflectance, 0.001f, 0.999f);
+            clamp_low_in_place(values->m_mfp, 1.0e-6f);
+
+            // Precompute scaling factor. 
+            for (size_t i = 0, e = Spectrum::size(); i < e; ++i)
+            {
+                const float albedo = albedo_from_reflectance(values->m_reflectance[i]);
+                const float s = normalized_diffusion_s_mfp(values->m_reflectance[i]);
+                const float extinction = rcp(values->m_mfp[i] * s);
+
+                values->m_volume_data.m_precomputed.m_extinction[i] = extinction;
+                values->m_volume_data.m_scattering[i] = albedo * extinction;
+                values->m_volume_data.m_absorption[i] = (1.0f - albedo) * extinction;
+            }
+        }
+
+        static float albedo_from_reflectance(const float r)
+        {
+            return 1.0f - exp(r * (-5.09406f + r * (2.61188f - 4.31805f * r)));
+        }
+
+        bool sample(
+            const ShadingContext&   shading_context,
+            SamplingContext&        sampling_context,
+            const void*             data,
+            const ShadingPoint&     outgoing_point,
+            const Vector3f&         outgoing_dir,
+            BSSRDFSample&           bssrdf_sample,
+            BSDFSample&             bsdf_sample) const override
+        {
+            const RandomWalkBSSRDFInputValues* values =
+                static_cast<const RandomWalkBSSRDFInputValues*>(data);
+
+            // Pick initial random-walk direction.
+            sampling_context.split_in_place(2, 1);
+            const Vector2d s = sampling_context.next2<Vector2d>();
+            Vector3d initial_dir = sample_hemisphere_cosine(s);
+            initial_dir.y = -initial_dir.y;
+            initial_dir = outgoing_point.get_shading_basis().transform_to_parent(initial_dir);
+            ShadingRay ray = ShadingRay(
```

Comment:
I think you're creating a copy here. Better and simpler:

```
ShadingRay ray(
    ...
    ...);
```

Also move closing parenthesis goes to the previous line.